### DISCUSSION TITLE
socket.io client always closing

### DISCUSSION THREAD

**Discussion by Notadumdum** [miguelgrinberg/python-socketio#991] [Q&A]
hey i tried making this connection work... i first installed the updated version of python-socketio[client]==5.7.1 but apparently this version supports EIO=4 and the socket.io client i'm trying to open is EIO=3.. so i had to uninstall it and install an old version python-socketio==4.6.1 . after that i tried running this code
```
import socketio

sio = socketio.Client(ssl_verify=False,logger=True,engineio_logger=True)

@sio.event
def connect():
    print("Connected!")
    sio.emit("authorization",{"session":"2ot3i7..................3h","isDemo":1})

@sio.event
def connect_error(data):
    print("The connection failed!\n\n" + data)

@sio.event
def disconnect():
    print("Disconnected!")


@sio.on('*')
def catch_all(event,data):
    pass



@sio.on('handshake')
def on_message(data):
    print('Handshake',data)
    sio.emit("instruments/update",{"asset":"GBPUSD","period":60})
    sio.emit("indicator/list")
    sio.emit("drawing/load")
    sio.emit("pending/list")
    sio.emit("chart_notification/get")
    sio.emit("depth/follow","GBPUSD")
    sio.emit("tick")
    sio.emit("settings/store",{"chartId":"graph","settings":{"chartId":"graph","chartType":2,"currentExpirationTime":0,"isFastOption":False,"isFastAmountOption":False,"isIndicatorsMinimized":False,"isIndicatorsShowing":True,"isShortBetElement":False,"chartPeriod":4,"currentAsset":{"symbol":"GBPUSD"},"dealValue":1000,"dealPercentValue":1,"isVisible":True,"timePeriod":60,"gridOpacity":8,"isAutoScrolling":1,"isOneClickTrade":True,"upColor":"#0FAF59","downColor":"#FF6251"}})
if __name__ == '__main__' :
    sio.connect('https://ws.domain.io/socket.io/', transports=['websocket'])
    sio.wait()
```
and here is what came back...

```
Attempting WebSocket connection to wss://ws.domain.io/socket.io/?transport=websocket&EIO=3
    WebSocket connection accepted with {'sid': '5AeKOf4oS0U-pHnaB1-2', 'upgrades': [], 'pingInterval': 25000, 'pingTimeout': 5000}
    Engine.IO connection established
    Sending packet PING data None
    Received packet MESSAGE data 0
    Received packet MESSAGE data 1
    Namespace / is connected
    Connected!
    Emitting event "authorization" [/]
    Sending packet MESSAGE data 2["authorization",{"session":"2ot......................3h","isDemo":1}]
    Disconnected!
    Received packet NOOP data None
    WebSocket connection was closed, aborting
    Waiting for write loop task to end
    Exiting write loop task
    Waiting for ping loop task to end
    PONG response has not been received, aborting
    Exiting ping task
    Engine.IO connection dropped
    Connection failed, new attempt in 1.41 seconds
    Exiting read loop task
    Attempting WebSocket connection to wss://ws.domain.io/socket.io/?transport=websocket&EIO=3
    WebSocket connection accepted with {'sid': 'sQvjGss3KsxnBzQ2B2gI', 'upgrades': [], 'pingInterval': 25000, 'pingTimeout': 5000}
    Engine.IO connection established
    Sending packet PING data None
    Reconnection successful
    Received packet MESSAGE data 0
    Namespace / is connected
    Received packet MESSAGE data 1
    Connected!
    Emitting event "authorization" [/]
    Received packet NOOP data None
    Sending packet MESSAGE data 2["authorization",{"session":"2ot.......................3h","isDemo":1}]
    WebSocket connection was closed, aborting
    Waiting for write loop task to end
    Exiting write loop task
    Waiting for ping loop task to end
    PONG response has not been received, aborting
    Exiting ping task
    Engine.IO connection dropped
    Disconnected!
    Disconnected!
    Connection failed, new attempt in 0.90 seconds
    Exiting read loop task
```

i had to kill the terminal cause it was trying to reconnect

please if you have any questions to make my explanations clear you can comment and i'll answer thank you if you can give an advice

**Comment by miguelgrinberg** [comment]
The log does not show anything abnormal. It seems the server disconnected you after you sent your authorization details.

**Reply by Notadumdum** [reply]
In the log.. 
```
Sending packet MESSAGE data 2["authorization",{"session":"2ot......................3h","isDemo":1}]
```
In devtools it shows 42 not just 2.. based on the engine.io number being 4 and the socket.io number being 2 in the payload... if that makes sense.. how would I go about this issue

**Reply by miguelgrinberg** [reply]
The 4 is the code for a packet of type `MESSAGE`, which is shown in text form. So as I said, this is all normal, no issues that I can see.

**Reply by Notadumdum** [reply]
So you're saying no problem with my code... and also the does the sending packet ```MESSAGE``` 2 mean ```42```?.. is that what you mean and also do you mean the server is the one closing the connection

**Reply by miguelgrinberg** [reply]
Yes to all.

**Reply by Notadumdum** [reply]
So is there nothing I can do??
Like maybe I'm sending something wrongly..will using flask or django change anything

**Reply by miguelgrinberg** [reply]
The best thing you can do is find out why the server is not accepting your auth. That is very likely the problem.

**Reply by Notadumdum** [reply]
Well I did a bit of studying and I didn't send in a "2" as it is shown in devtools.. I don't know if it will work though but I'll just keep on looking for solutions..maybe then I'll close the question

**Reply by miguelgrinberg** [reply]
You are looking at this the wrong way. The "4" and the "2" are internal to the Socket.IO and Engine.IO protocols. You never send those. If you aren't familiar with these protocols you will not be able to make sense of a byte-by-byte dump of what is sent.

As I said many times above, you continue to ignore the most likely cause of your issue, which is that your authentication is bad.

**Reply by Notadumdum** [reply]
Hey Miguel Grinberg.. yesterday I was looking through this website's certificate and I saw a part that says 
```EXTENDED KEY USAGES----(Purposes): Server Authentication,Client Authentication```
So the question is do you think the connection I'm trying to authenticate needs the ssl certificate for it to work
![IMG-20220824-WA0017](https://user-images.githubusercontent.com/105162586/186394550-c1278db8-12ae-4ae9-93f5-f124e7562cd1.jpg)
![IMG-20220824-WA0016](https://user-images.githubusercontent.com/105162586/186394557-47d92463-cabb-401a-a7fe-6bab8097af77.jpg)
![IMG-20220824-WA0015](https://user-images.githubusercontent.com/105162586/186394562-09ff7b4f-abfa-4fa6-9672-8474f5f582ca.jpg)

**Reply by miguelgrinberg** [reply]
I think you may be confusing user authentication with SSL certificate validation and encryption, which are completely different things. SSL certificates are not handled by this library at all, this is something that is handled between the HTTP client and the web server. So I really don't know how to answer your question, to me it seems completely unrelated to the topic of Socket.IO.

**Reply by Notadumdum** [reply]
Ok

**Reply by Notadumdum** [reply]
Have you ever received any complaints as close to mine and the person has fixed it.. if so can you please send a link or something.. I'm really clueless

**Reply by miguelgrinberg** [reply]
I mentioned this a few times already. The logs that you shared do not suggest there are any issues. You are passing some sort of token to the server right after you connect:

    sio.emit("authorization",{"session":"2ot3i7..................3h","isDemo":1})

The server does not appear to like it and is ending the connection after this.

The most likely issue here is that this session value that you are passing is not being accepted by the server.
